std/base.m31

```diff
@@ -12,7 +12,7 @@ handle
   | as_eq _ => None
 end
 
-let (|>) x f = f x
+let (|>) x f : Π α β, α → (α → β) → β = f x
```

Comment:
In this example we compute
`fun x f => f x : a -> b -> c` with side condition `AppConstraint(b, a, c)`
Since `a`, `b` and `c` appear in the environment they can't be generalised, so we bind `(|>)` with that type.
Then at some point we apply it to some list (in `list_map` which never gets used so the list type is `list alpha`) so we instantiate `a` to `list alpha`, `b` to `list alpha -> list alpha` and the condition resolves to `c = list alpha` (or something like that, not sure of the order of instantiations).
We only figure out that the side condition is resolved after we instantiate the type, so we would have to do some crazy backtracking to generalize properly.

I thought about changing schemas so they can generalize side conditions as well, so we would get
`(|>) : forall a b c [App(b, a, c)], a -> b -> c`
Then when we instantiate a schema we would add a bunch of side conditions to the environment.
It would allow applying `(|>)` to both judgements and functions, which could be cool or confusing. The types we get are harder to read, but that might just be a printing problem.